README
HTML::Tidy
==========

HTML::Tidy is an HTML checker in a handy dandy object.  It's meant as
a replacement for HTML::Lint.

INSTALLATION

You need libtidy to build HTML::Tidy. Check section INSTALLING TIDY on this
document for more information.

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

INSTALLING TIDY

You can get a distribution for libtidy at:

    http://tidy.sourceforge.net/src/tidy_src.tgz

You can also check out from the CVS tree with the following two steps:

Login in to the anonymous CVS server at SourceForge:

    $ cvs -d:pserver:anonymous@cvs.sf.net:/cvsroot/tidy login
    (Logging in to anonymous@cvs.sf.net)
    CVS password: [hit ENTER]

Check out the source tree:

    $ cvs -d:pserver:anonymous@cvs.sf.net:/cvsroot/tidy co tidy
    cvs server: Updating tidy
    cvs server: Updating tidy/CVSROOT
    U tidy/CVSROOT/checkoutlist
    U tidy/CVSROOT/commitinfo
    ... many many files come out ...

If you're on Mac OS X, you'll need to tell the Makefile that you use
ranlib:

    $ export set RANLIB=ranlib

Change to the directory with the Makefile in it, and run make.
This example uses the GNU make Makefile.

    $ cd tidy/build/gmake/

    $ make
    if [ ! -d ./obj ]; then mkdir ./obj; fi
    gcc -o obj/access.o ...
    ... etc etc etc ...

Install the libs, headers and the tidy executable:

    $ sudo make install

If you're on Mac OS X, you'll have to run ranlib again on the installed
lib:

    $ sudo ranlib /usr/local/lib/libtidy.a

Once these steps are complete, you'll be able to go back to HTML::Tidy
and build it normally.


COPYRIGHT AND LICENCE

Copyright (C) 2004 by Andy Lester

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.1 or,
at your option, any later version of Perl 5 you may have available.


Changes
Revision history for Perl extension HTML::Tidy.

1.06    Fri Aug 19 10:23:29 CDT 2005
        NO FUNCTIONAL CHANGES.  Just a version bump and some doc fixes.

1.05_02 Wed Apr 20 22:43:31 CDT 2005
        [FIXES]
        * Made clean() return undef if it has no input.  Thanks to
          Dominique Quatravax.  See RT #8051.
        * Removed line-endings so that we're happy cross-platform in
          the tests.  Thanks to the anonymous contributor of RT #7836.

        [ENHANCEMENTS]
        * Added support for config files.  Thanks to Anders Ardo!
          See RT #11120.

1.05_01 Thu Sep  2 21:01:52 CDT 2004
        [FIXES]
        * Fixed some segfaults when calling $tidy->clean().  Thanks
          to Aaron Patterson, fixes CPAN #7254.

        * Made the Makefile.PL more portable under Mac OS X where
          libtidy had been installed by Fink.  Also added some
          Debian and OS X notes to the README.  Thanks to Dan
          Friedman.

1.04    Wed May 12 15:18:37 CDT 2004
        [FIXES]
        * If for some reason, Tidy can't create the cleaned up doc,
          we don't segfault.  Added a test to check for it with an HTML
          set that we know made it segfault in 1.02.

        [ENHANCEMENTS]
        * Portability enhancements from Dave Rolsky.

1.02    Wed Mar  3 22:43:57 CST 2004
        [ENHANCEMENTS]
        * Added clean() method for cleaning up using the Tidy magic.
          Thanks to Ben Trott.

        [FIXES]
        * Message parsing is split on a Ctrl-J instead of \n now.
          This should fix the test problems on Win32.

        * Removed the #include "ppport.h" from Tidy.xs so it can compile
          on Win32.  Thanks to crazyinsomniac.

        [MISCELLANEOUS]
        * Moved the install directions for tidylib into the README.

1.00    Wed Feb 25 22:19:28 CST 2004

        First real version, released on an unsuspecting world.

        If you were one of the ones to get in on 0.01 thru 0.04, you may
        want to check your code.  The API has changed some.  I know that
        parse_file() is now parse(), for example.  There may be others.

        Going forward, of course, I'll note any other API changes here.


lib/HTML/Tidy.pm
package HTML::Tidy;

use 5.006001;
use strict;
use warnings;

use HTML::Tidy::Message;

=head1 NAME

HTML::Tidy - Web validation in a Perl object

=head1 VERSION

Version 1.06

=cut

our $VERSION = "1.06";

=head1 SYNOPSIS

    use HTML::Tidy;

    my $tidy = HTML::Tidy->new( {config_file => 'path/to/config'} );
    $tidy->ignore( type => TIDY_WARNING );
    $tidy->parse( "foo.html", $contents_of_foo );

    for my $message ( $tidy->messages ) {
        print $message->as_string;
    }

=head1 Description

C<HTML::Tidy> is an HTML checker in a handy dandy object.  It's meant as
a replacement for L<HTML::Lint>.  If you're currently an L<HTML::Lint>
user looking to migrate, see the section L</Converting from HTML::Lint>.

=head1 Exports

Message types C<TIDY_WARNING> and C<TIDY_ERROR>.

Everything else is an object method.

=cut

require Exporter;

our @ISA = qw( Exporter DynaLoader );

use constant TIDY_ERROR => 2;
use constant TIDY_WARNING => 1;

our @EXPORT = qw( TIDY_ERROR TIDY_WARNING );

=head1 Methods

=head2 new()

Create an HTML::Lint object.

    my $tidy = HTML::Tidy->new();

Optionally you can give a hashref of configuration parms.  Currently,
only C<config_file> is supported.

    my $tidy = HTML::Tidy->new( {config_file => 'path/to/tidy.cfg'} );

This configuration file will be read and used when you clean an HTML file.

=cut

sub new {
    my $class = shift;
    my $args = shift || {};

    my $self = bless {
        messages => [],
        ignore_type => [],
        ignore_text => [],
        config_file => "",
    }, $class;

    for my $key ( qw( config_file ) ) {
        if ( exists $args->{$key} ) {
            $self->{$key} = $args->{$key};
        }
    }

    return $self;
}

=head2 messages()

Returns the messages accumulated.

=cut

sub messages {
    my $self = shift;

    return @{$self->{messages}};
}

=head2 clear_messages()

Clears the list of messages, in case you want to print and clear, print
and clear.  If you don't clear the messages, then each time you call
L<parse()> you'll be accumulating more in the list.

=cut

sub clear_messages {
    my $self = shift;

    $self->{messages} = [];
}

=head2 ignore( parm => value [, parm => value ] )

Specify types of messages to ignore.  Note that the ignore flags must be
set B<before> calling C<parse()>.  You can call C<ignore()> as many times
as necessary to set up all your restrictions; the options will stack up.

=over 4

=item * type => TIDY_(WARNING|ERROR)

Specifies the type of messages you want to ignore, either warnings
or errors.  If you wanted, you could call ignore on both and get no
messages at all.

    $tidy->ignore( type => TIDY_WARNING );

=item * text => qr/regex/

=item * text => [ qr/regex1/, qr/regex2/, ... ]

Checks the text of the message against the specified regex or regexes,
and ignores the message if there's a match.  The value for the I<text>
parm may be either a regex, or a reference to a list of regexes.

    $tidy->ignore( text => qr/DOCTYPE/ );
    $tidy->ignore( text => [ qr/unsupported/, qr/proprietary/i ] );

=back

=cut

sub ignore {
    my $self = shift;
    my @parms = @_;

    while ( @parms ) {
        my $parm = shift @parms;
        my $value = shift @parms;
        my @values = ref($value) eq "ARRAY" ? @$value : ($value);

        die "Invalid ignore type of \"$parm\"" unless ($parm eq "text") or ($parm eq "type");

        push( @{$self->{"ignore_$parm"}}, @values );
    } # while
} # ignore

=head2 parse( $filename, $str [, $str...] )

Parses a string, or list of strings, that make up a single HTML file.

The I<$filename> parm is only used as an identifier for your use.
The file is not actually read and opened.

Returns true if all went OK, or false if there was some problem calling
tidy, or parsing tidy's output.

=cut

sub parse {
    my $self = shift;
    my $filename = shift;

    my $html = join( "", @_ );

    my $errorblock = _tidy_messages( $html );
    return unless defined $errorblock;


    return !$self->_parse_errors($filename, $errorblock);
}

sub _parse_errors {
    my $self = shift;
    my $filename = shift;
    my $errs = shift;

    my $parse_errors;

    my @lines = split( /\012/, $errs );
    for my $line ( @lines ) {
        chomp $line;

        my $message;
        if ( $line =~ /^line (\d+) column (\d+) - (Warning|Error): (.+)$/ ) {
            my $type = ($3 eq "Warning") ? TIDY_WARNING : TIDY_ERROR;
            $message = HTML::Tidy::Message->new( $filename, $type, $1, $2, $4 );

        }
        elsif ( $line =~ /^\d+ warnings?, \d+ errors? were found!/ ) {
            # Summary line we don't want

        }
        elsif ( $line eq "No warnings or errors were found." ) {
            # Summary line we don't want

        }
        elsif ( $line eq "This document has errors that must be fixed before" ){
            # Summary line we don't want

        }
        elsif ( $line eq "using HTML Tidy to generate a tidied up version." ){
            # Summary line we don't want

        }
        elsif ( $line =~ m/^Info:/  ) {
            # Info line we don't want

        }
        elsif ( $line =~ m/^\s*$/  ) {
            # Blank line we don't want

        }
        else {
            warn "Unknown error type: $line";
            ++$parse_errors;
        }
        push( @{$self->{messages}}, $message ) if $message && $self->_is_keeper( $message );
    } # for
    return $parse_errors;
}

=head2 clean( $str [, $str...] )

Cleans a string, or list of strings, that make up a single HTML file.

Returns true if all went OK, or false if there was some problem calling
tidy, or parsing tidy's output.

=cut

sub clean {
    my $self = shift;

    my $text = join( "", @_ );
    if ( defined $text ) {
        $text .= "\n";
    }
    my ($cleaned, $errbuf) = _tidy_clean( $text, $self->{config_file} );

    $self->_parse_errors('', $errbuf);
    return $cleaned;
}


# Tells whether a given message object is one that we should keep.

sub _is_keeper {
    my $self = shift;

    my $message = shift;

    my @ignore_types = @{$self->{ignore_type}};
    if ( @ignore_types ) {
        my $type = $message->type;
        return if grep { $type == $_ } @ignore_types;
    }

    my @ignore_texts = @{$self->{ignore_text}};
    if ( @ignore_texts ) {
        my $text = $message->text;
        return if grep { $text =~ $_ } @ignore_texts;
    }

    return 1;
}

require XSLoader;
XSLoader::load('HTML::Tidy', $VERSION);

1;

__END__

=head1 Converting From HTML::Lint

L<HTML::Tidy> is different from L<HTML::Lint> in a number of crucial ways.

=over 4

=item * It's not pure Perl

C<HTML::Tidy> is mostly a happy wrapper around libtidy.

=item * The real work is done by someone else

Changes to libtidy may come down the pipe that I don't have control over.
That's the price we pay for having it do a darn good job.

=item * It's no longer bundled with its C<Test::> counterpart

L<HTML::Lint> came bundled with C<Test::HTML::Lint>, but
L<Test::HTML::Tidy> is a separate distribution.  This saves the people
who don't want the C<Test::> framework from pulling it in, and all its
prerequisite modules.

=back

=head1 Bugs & Feedback

I welcome your comments and suggestions.  Please send them to
C<< <bug-html-tidy at rt.cpan.org> >> so that they can be tracked in the
RT ticket tracking system.

=head1 Author

Andy Lester, C<< <andy at petdance.com> >>

=head1 Copyright & License

Copyright (C) 2005 by Andy Lester

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.1 or,
at your option, any later version of Perl 5 you may have available.

=cut


lib/HTML/Tidy/Message.pm
# $Id: Message.pm 65 2004-02-25 06:22:56Z andy $
package HTML::Tidy::Message;

use strict;

=head1 NAME

HTML::Tidy::Message - Message object for the Tidy functionality

=head1 SYNOPSIS

See L<HTML::Tidy> for all the gory details.

=head1 EXPORTS

None.  It's all object-based.

=head1 METHODS

Almost everything is an accessor.

=head2 new( $file, $line, $column, $text )

Create an object.  It's not very exciting.

=cut

sub new {
    my $class = shift;

    my $file = shift;
    my $type = shift;
    my $line = shift || 0;
    my $column = shift || 0;
    my $text = shift;

    # Add an element that says what tag caused the error (B, TR, etc)
    # so that we can match 'em up down the road.
    my $self  = {
        _file => $file,
        _type => $type,
        _line => $line,
        _column => $column,
        _text => $text,
    };

    bless $self, $class;

    return $self;
}

=head2 where()

Returns a formatted string that describes where in the file the
error has occurred.

For example,

    (14:23)

for line 14, column 23.

The terrible thing about this function is that it's both a plain
ol' formatting function as in

    my $str = where( 14, 23 );

AND it's an object method, as in:

    my $str = $error->where();

I don't know what I was thinking when I set it up this way, but
it's bad practice.

=cut

sub where {
    my $self = shift;

    return "-" unless $self->line && $self->column;

    return sprintf( "(%d:%d)", $self->line, $self->column );
}

=head2 as_string()

Returns a nicely-formatted string for printing out to stdout or some similar user thing.

=cut

sub as_string {
    my $self = shift;

    my %strings = (
        1 => "Warning",
        2 => "Error",
    );

    return sprintf( "%s %s %s: %s", $self->file, $self->where, $strings{$self->type}, $self->text );
}

=head2 file()

Returns the filename of the error, as set by the caller.

=head2 type()

Returns the type of the error.  This will either be C<TIDY_ERROR>,
or C<TIDY_WARNING>.

=head2 line()

Returns the line number of the error, or 0 if there isn't an applicable
line number.

=head2 column()

Returns the column number, or 0 if there isn't an applicable column
number.

=head2 text()

Returns the text of the message.  This does not include a type string,
like "Info: ".

=cut

sub file    { my $self = shift; return $self->{_file} }
sub type    { my $self = shift; return $self->{_type} }
sub line    { my $self = shift; return $self->{_line} }
sub column  { my $self = shift; return $self->{_column} }
sub text    { my $self = shift; return $self->{_text} }


=head1 LICENSE

This code may be distributed under the same terms as Perl itself.

Please note that these modules are not products of or supported by the
employers of the various contributors to the code.

=head1 AUTHOR

Andy Lester, C<< <andy@petdance.com> >>

=cut

1; # happy
